### PR TITLE
RavenDB-18852 Support range queries using the index instead of scan

### PR DESCRIPTION
**Pull request by ayende**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-18852 

### Additional description

This PR adds support for running range queries by using `MultiTermProvider`.
We scan through the index and do efficient range queries to provide the relevant terms.

I provided only the base infrastructure, still need to wire it in all the right places. 

Also added support for storing `FixedSizeTree` where the key is a `double`. 

### Type of change

- New feature

### How risky is the change?

- Not relevant

### Backward compatibility

- Not relevant

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- Tests have been added that prove the fix is effective or that the feature works

### Is there any existing behavior change of other features due to this change?

Need to wire that to the Corax queries execution

### UI work

- No UI work is needed

### CODE REVIEW

File: bench/Micro.Benchmark/Micro.Benchmark.csproj [hunk 1]
```diff
@@ -4,6 +4,8 @@
     <RuntimeFrameworkVersion>6.0.6</RuntimeFrameworkVersion>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
     <AssemblyName>Micro.Benchmark</AssemblyName>
+    <EnablePreviewFeatures>True</EnablePreviewFeatures>
```
Comment (ayende) [inline bench/Micro.Benchmark/Micro.Benchmark.csproj:7]: I'm using `Generic Math` here, so we have to use the preview feature.
@arek / @ppekrol - I don't _expect_ that to be an issue, but wanted your eyes on that as well.

File: src/Corax/IndexFieldsMapping.cs [hunk 2]
```diff
@@ -30,11 +30,37 @@ public IndexFieldsMapping(ByteStringContext context)
         _fieldsList = new List<IndexFieldBinding>();
     }
 
+    
+    private const int I64Postfix = 875972909; //"-I64"
+    private const int DblPostfix = 1818379309; //"-Dbl"
```
Comment (ayende) [inline src/Corax/IndexFieldsMapping.cs]: This is a bit different from what we do for Lucene. Where the suffixes are `_L` and `_D`. I don't think it matters, but wanted to point it out.

File: src/Corax/IndexFieldsMapping.cs [hunk 3]
```diff
@@ -30,11 +30,37 @@ public IndexFieldsMapping(ByteStringContext context)
...
+        return scope;
+    }
+
     public IndexFieldsMapping AddBinding(int fieldId, Slice fieldName, Analyzer analyzer = null, bool hasSuggestion = false, FieldIndexingMode fieldIndexingMode = FieldIndexingMode.Normal, bool hasSpatial = false)
     {
         if (!_fieldsById.TryGetValue(fieldId, out var storedAnalyzer))
         {
-            var binding = new IndexFieldBinding(fieldId, fieldName, analyzer, hasSuggestion, fieldIndexingMode, hasSpatial);
+            GetFieldNameForDoubles(fieldName, out var fieldNameDouble);
+            GetFieldNameForLongs(fieldName, out var fieldNameLong);
```
Comment (ayende) [inline src/Corax/IndexFieldsMapping.cs:61]: @maciejaszyk Here we _have_ to check that the lifetime of the `IndexFeidlsMapping` is as long as the `context`.
I'm fairly certain that it is, but I want another pair of eyes on that.

File: src/Corax/IndexSearcher/IndexSearcher.MultiTermMatches.cs [hunk 4]
```diff
@@ -321,6 +366,96 @@ public AndNotMatch NotInQuery<TInner>(string field, TInner inner, List<string> n
...
+            // (true, NullScoreFunction) => MultiTermMatch.Create(new MultiTermMatch<NotStartWithTermProvider>(_transaction.Allocator,
+            //     new NotStartWithTermProvider(this, _transaction.Allocator, terms, fieldName, fieldId, slicedTerm))),
+            //
+            // (false, _) => MultiTermMatch.Create(
+            //     MultiTermBoostingMatch<StartWithTermProvider>.Create(
+            //         this, new StartWithTermProvider(this, _transaction.Allocator, terms, fieldName, fieldId, slicedTerm), scoreFunction)),
+            //
+            // (true, _) => MultiTermMatch.Create(
+            //     MultiTermBoostingMatch<NotStartWithTermProvider>.Create(
+            //         this, new NotStartWithTermProvider(this, _transaction.Allocator, terms, fieldName, fieldId, slicedTerm), scoreFunction))
```
Comment (ayende) [inline src/Corax/IndexSearcher/IndexSearcher.MultiTermMatches.cs:391]: Need to fill this in.

File: src/Corax/IndexSearcher/IndexSearcher.MultiTermMatches.cs [hunk 5]
```diff
@@ -321,6 +366,96 @@ public AndNotMatch NotInQuery<TInner>(string field, TInner inner, List<string> n
...
+            return MultiTermMatch.CreateEmpty(_transaction.Allocator);
+
+        var set = fields?.FixedTreeFor(fieldLong, sizeof(long));
+
+        return (isNegated, scoreFunction) switch
+        {
+            (false, NullScoreFunction) => MultiTermMatch.Create(new MultiTermMatch<TermNumericRangeProvider<TLow, THigh, long>>(_transaction.Allocator,
+                new TermNumericRangeProvider<TLow, THigh, long>(this, set, terms, fieldName, low, high))),
+
+            _ => throw new NotSupportedException()
```
Comment (ayende) [inline src/Corax/IndexSearcher/IndexSearcher.MultiTermMatches.cs:412]: Need to fill this in.

File: src/Corax/IndexSearcher/IndexSearcher.MultiTermMatches.cs [hunk 6]
```diff
@@ -321,6 +366,96 @@ public AndNotMatch NotInQuery<TInner>(string field, TInner inner, List<string> n
...
+            return MultiTermMatch.CreateEmpty(_transaction.Allocator);
+
+        var set = fields?.FixedTreeForDouble(fieldLong, sizeof(long));
+
+        return (isNegated, scoreFunction) switch
+        {
+            (false, NullScoreFunction) => MultiTermMatch.Create(new MultiTermMatch<TermNumericRangeProvider<TLow, THigh, double>>(_transaction.Allocator,
+                new TermNumericRangeProvider<TLow, THigh, double>(this, set, terms, fieldName, low, high))),
+
+            _ => throw new NotSupportedException()
```
Comment (ayende) [inline src/Corax/IndexSearcher/IndexSearcher.MultiTermMatches.cs:443]: Need to fill this in.

File: src/Corax/IndexWriter.cs [hunk 7]
```diff
@@ -84,8 +87,14 @@ private IndexWriter(IndexFieldsMapping fieldsMapping)
...
             var bufferSize = fieldsMapping!.Count;
             _buffer = new Dictionary<Slice, List<long>>[bufferSize];
+            _bufferDoubles = new Dictionary<double, List<long>>[bufferSize];
+            _bufferLongs = new Dictionary<long, List<long>>[bufferSize];
             for (int i = 0; i < bufferSize; ++i)
+            {
                 _buffer[i] = new Dictionary<Slice, List<long>>(SliceComparer.Instance);
+                _bufferDoubles[i] = new Dictionary<double, List<long>>();
+                _bufferLongs[i] = new Dictionary<long, List<long>>();
+            }
```
Comment (ayende) [inline src/Corax/IndexWriter.cs:93]: we should probably add a cache here for those, no?

File: src/Corax/Queries/TermProviders/TermProvider.Range.cs [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+    {
+        private readonly IndexSearcher _searcher;
+        private readonly Slice _fieldName;
+        private readonly TVal _low, _high;
+        private readonly CompactTree _tree;
+        private readonly FixedSizeTree<TVal> _set;
+        private readonly FixedSizeTree<TVal>.IFixedSizeIterator _iterator;
+        private bool _first;
+
+        private const int TermBufferSize = 32;
```
Comment (ayende) [inline src/Corax/Queries/TermProviders/TermProvider.Range.cs:131]: `long.MaxValue` len in chars == 19
`double.MaxValue` len in chars == 21

File: src/Corax/Queries/UnaryMatch.Ordinal.cs [hunk 9]
```diff
@@ -947,7 +947,7 @@ private static int FillFuncNumericalAll<TComparer>(ref UnaryMatch<TInner, TValue
                 return new UnaryMatch<TInner, TValueType>(
                     in inner, UnaryMatchOperation.GreaterThan,
                     searcher, fieldId, value,
-                    mode == UnaryMatchOperationMode.Any ? &FillFuncNumericalAny<GreaterThanMatchComparer> : &FillFuncNumericalAny<GreaterThanMatchComparer>, &AndWith,
+                    mode == UnaryMatchOperationMode.Any ? &FillFuncNumericalAny<GreaterThanMatchComparer> : &FillFuncSequenceAll<GreaterThanMatchComparer>, &AndWith,
```
Comment (ayende) [inline src/Corax/Queries/UnaryMatch.Ordinal.cs]: Pretty sure that this is a valid fix, but I want another pair of eyes here.

File: src/Corax/Queries/TermProviders/TermProvider.Range.cs [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+        }
+
+        public string DebugView => Inspect().ToString();
+    }
+    
+     [DebuggerDisplay("{DebugView,nq}")]
+    public unsafe struct TermNumericRangeProvider<TLow, THigh, TVal> : ITermProvider
+        where TLow : struct, Range.Marker
+        where THigh  : struct, Range.Marker
+        where TVal : unmanaged, IBinaryNumber<TVal>, IMinMaxValue<TVal>, INumber<TVal>
```
Comment (ayende) [inline src/Corax/Queries/TermProviders/TermProvider.Range.cs:121]: This is where we use generic math

File: src/Corax/IndexFieldsMapping.cs (hunk 2 above)
Comment (ayende) [reply src/Corax/IndexFieldsMapping.cs]: Changed that to match Lucene.

File: src/Corax/Queries/UnaryMatch.Ordinal.cs (hunk 9 above)
Comment (maciejaszyk) [reply src/Corax/Queries/UnaryMatch.Ordinal.cs]: This is the builder for numerics. Notice the above condition when we check if `TValueType` is a `Slice`.

Comment (ayende) [reply src/Corax/Queries/UnaryMatch.Ordinal.cs]: So this should be `FillFuncNumericalAll` ?

Comment (maciejaszyk) [reply src/Corax/Queries/UnaryMatch.Ordinal.cs]: I think so

File: src/Corax/IndexWriter.cs [hunk 11]
```diff
@@ -244,10 +254,11 @@ void AddMaybeAvoidDuplicate(List<long> term, long entryId)
                     break;
 
                 case IndexEntryFieldType.Tuple:
-                    if (entryReader.Read(binding.FieldId, out Span<byte> valueInEntry) == false)
+                    if (entryReader.Read(binding.FieldId, out _,  out long lVal, out double dVal,out Span<byte> valueInEntry) == false)
                         break;
 
                     ExactInsert(valueInEntry);
+                    NumericInsert(lVal, dVal);
                     break;
```
Comment (redknightlois) [inline src/Corax/IndexWriter.cs:256]: Try `NumericInsert` on top instead, looks like something with the parser.

File: src/Corax/IndexWriter.cs [hunk 12]
```diff
@@ -677,25 +649,144 @@ public void Commit()
...
+                entries.Add(existing);
+                return AddNewTerm(entries, tmpBuf, out  termId);
+            }
+        }
+
+        private unsafe void InsertNumericFieldLongs(Tree fieldsTree, int fieldId, Span<byte> tmpBuf)
+        {
+            FixedSizeTree fieldTree = fieldsTree.FixedTreeFor(_fieldsMapping.GetByFieldId(fieldId).FieldNameLong, sizeof(long));
+            var llt = Transaction.LowLevelTransaction;
+            var sortedTerms = _bufferLongs[fieldId].Keys.ToArray();// TODO: let's avoid this allocation
```
Comment (redknightlois) [inline src/Corax/IndexWriter.cs:782]: We haven't done allocations optimizations for the most part. Dont worry about it.

File: src/Corax/Queries/UnaryMatch.Ordinal.cs (hunk 9 above)
Comment (redknightlois) [reply src/Corax/Queries/UnaryMatch.Ordinal.cs]: Yes.

File: src/Corax/IndexWriter.cs (hunk 7 above)
Comment (redknightlois) [reply src/Corax/IndexWriter.cs:93]: Yes, but we havent applied allocations optimizations yet. We only applied for the most blatant offenders. There is lots of opportunities to improve that, and this is just one of those. I wouldnt worry much for now.

File: src/Corax/IndexWriter.cs (hunk 11 above)
Comment (ayende) [reply src/Corax/IndexWriter.cs:256]: I don't follow? 
This insert a numeric token

**Comment by redknightlois** [comment]
Can be closed now. Will be merged from https://github.com/ravendb/ravendb/pull/14492

**Pull request closed without merging**